syntax = "proto3";

package jobplatform;

// -------- this is python side services --------
service ProfileAIService {
  rpc GenerateGitHubSummary (GitHubSummaryRequest) returns (GitHubSummaryResponse);
  rpc UploadResume (ResumeUploadRequest) returns (ResumeUploadResponse);
  rpc UpdateResume (UpdateResumeRequest) returns (ResumeResponse);
  rpc GenerateResume (GenerateResumeRequest) returns (ResumeResponse);
}

service PracticeAIService {
  rpc PracticeQuestion (PracticeQuestionRequest) returns (PracticeQuestionResponse);
}

service VectorAIService {
  rpc GetEmbedding (EmbeddingRequest) returns (EmbeddingResponse);
  rpc Similarity (SimilarityRequest) returns (SimilarityResponse);
  rpc FastSimilarity (SimilarityRequest) returns (FastSimilarityResponse);
}

// this is flow from node to python
message GitHubSummaryRequest {
  repeated string usernames = 1;
}

message GitHubSummaryResponse {
  string overall_summary = 1;
  float overall_rating = 2;
  repeated string overall_tips = 3;
  repeated string top_languages = 4;
}

message ResumeUploadRequest {
  bytes file_content = 1;
  string filename = 2;
}

message ResumeUploadResponse {
  string extracted_text = 1;
}

message UpdateResumeRequest {
  string existing_resume = 1;
  string new_details = 2;
}

message GenerateResumeRequest {
  string name = 1;
  repeated string skills = 2;
  string summary = 3;
}

message ResumeResponse {
  string name = 1;
  string summary = 2;
  repeated string skills = 3;
  string json_resume = 4; 
} 

message PracticeQuestionRequest {
  string user_id = 1;
  string prompt = 2;
  string job_description = 3;
  string practice = 4;
}

message PracticeQuestionResponse {
  string success = 1;
  // later to declare datatype
}

message EmbeddingRequest {
  string text = 1;
}

message EmbeddingResponse {
}

message SimilarityRequest {
  string text1 = 1;
  string text2 = 2; 
}

message SimilarityResponse {
  float embedding_similarity = 1;
  float fuzz_similarity = 2;
  float combined_similarity = 3;
  string canonical1 = 4;
  string canonical2 = 5;
}

 message FastSimilarityResponse {
  float similarity = 1;
  string canonical1 = 2;
  string canonical2 = 3;
 } 

// -------- Node-side platform services --------
// (future, not required now)
service JobPlatformService {
  rpc GetUserProfile (UserProfileRequest) returns (UserProfileResponse);
  rpc GetJob (JobRequest) returns (JobResponse);
}

message UserProfileRequest {
  int32 userId = 1;
}

message UserProfileResponse {
  int32 userId = 1;
  string fullName = 2;
  string professionalRole = 3;
  string companyName = 4;
}

message JobRequest {
  int32 jobId = 1;
}

message JobResponse {
  int32 jobId = 1;
  string title = 2;
  string description = 3;
  string companyName = 4;
}
